pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -601,29 +623,28 @@ func (r *ReconcileMachinePool) syncMachineSets(
 				}
 
 				// Carry over the taints on the remote machineset from the generated machineset. Merging strategy preserves the unique taints of remote machineset.
-				if rMS.Spec.Template.Spec.Taints == nil {
-					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
-				}
-				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByKey := make(map[string]*corev1.Taint)
-				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByKey[gTaint.Key] = &ms.Spec.Template.Spec.Taints[gIndex]
-				}
-				// Go through the remote MachineSet's taints, replacing overlaps with the generated MachineSet
-				for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
-					if gTaint, foundTaint := gTaintsByKey[rTaint.Key]; foundTaint {
-						if gTaint.Value != rTaint.Value || gTaint.Effect != rTaint.Effect {
-							rMS.Spec.Template.Spec.Taints[rIndex] = *gTaint
-							objectModified = true
-						}
-						delete(gTaintsByKey, rTaint.Key)
+				// First, build a map of remote MachineSet taints, so we simultaneously collapse duplicate entries.
+				taintMap := make(map[hivev1.TaintIdentifier]corev1.Taint)
+				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
+					if taintsToDelete.Has(controllerutils.IdentifierForTaint(&rTaint)) {
+						// This entry would be deleted from the final list
+						objectModified = true
+						continue
+					}
+					// Preserve the taint entry first encountered.
+					if !controllerutils.TaintExists(taintMap, &rTaint) {
+						taintMap[controllerutils.IdentifierForTaint(&rTaint)] = rTaint
```

Comment:
Hm, I think if we *don't* add this entry -- because it's a dup -- we need to set `objectModified`.
```suggestion
						taintMap[controllerutils.IdentifierForTaint(&rTaint)] = rTaint
					} else {
						// Skip this taint, which effectively means we're removing a duplicate.
						objectModified = true
```

We should have UT for this path: I.e. when the only "modification" is removing a dup from the rMS, make sure we're doing an `Update()`. As currently written, I don't think we will.